### PR TITLE
SliceOp verifier fix

### PR DESCRIPTION
**Pull request by mmanzoorTT**
### Ticket
closes https://github.com/tenstorrent/tt-mlir/issues/4210

### Problem description
TTIR/TTNN verifiers consider SliceOp illegal if start/end are outside of the [-dimSize, dimSize)
However the following PyTorch code is a legal op and generates start/end outside of range.
```
x = torch.randn((10, 3, 64, 128), dtype=torch.float32)
x = x[::3, ::3, ::4, 128:128:8]
```

### What's changed
Update TTIR/TTNN verifiers to not consider such ops as illegal.

### Checklist
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by mmanzoorTT** [comment]
> This is an interesting case that could have far-reaching consequences beyond slice op. Can you give some context where this appears? What's the semantics of a tensor with one of it's axis equal to 0? Does `15x20x0xType` have the same type (or to be more precise, equivalent) as `0x100x1000xType` (i.e. the same as canonical `0xType`)? If tensor is dense it's buffer size should be equal to the volume of that tensor, so if we go by that, answer on the previous question should be yes.

**Background**
We are getting this case for `microsoft/Phi-3-mini-128k-instruct` in [tt-torch test](https://github.com/tenstorrent/tt-torch/blob/main/tests/models/phi/test_phi_3.py).

relevant torch mlir graph snippet is
```
# File: tt-torch/env/venv/lib/python3.10/site-packages/transformers/models/phi3/modeling_phi3.py:188 in forward, code: query_states = query_states.view(hidden_shape).transpose(1, 2)
view_7: "f32[1, 29, 32, 96]" = torch.ops.aten.view.default(slice_19, [1, 29, -1, 96]);  slice_19 = None
permute_2: "f32[1, 32, 29, 96]" = torch.ops.aten.permute.default(view_7, [0, 2, 1, 3]);  view_7 = None

# File: tt-torch/env/venv/lib/python3.10/site-packages/transformers/models/phi3/modeling_phi3.py:144 in apply_rotary_pos_emb, code: q_rot, q_pass = q[..., :rotary_dim], q[..., rotary_dim:]
slice_22: "f32[1, 32, 29, 0]" = torch.ops.aten.slice.Tensor(permute_2, 3, 96, 9223372036854775807);  permute_2 = None

# File: tt-torch/env/venv/lib/python3.10/site-packages/transformers/models/phi3/modeling_phi3.py:147 in apply_rotary_pos_emb, code: q_embed = torch.cat([(q_rot * cos) + (rotate_half(q_rot) * sin), q_pass], dim=-1)
add_2: "f32[1, 32, 29, 96]" = torch.ops.aten.add.Tensor(mul_5, mul_6);  mul_5 = mul_6 = None
cat_2: "f32[1, 32, 29, 96]" = torch.ops.aten.cat.default([add_2, slice_22], -1);  add_2 = slice_22 = None
```
It is part of huggingface transformer package. You can find the implementation [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/phi3/modeling_phi3.py). The line numbers are different due to different version.

**Empty Tensor semantics**
This will create an empty tensor (with `numel() == 0`) and I believe that the buffer size will be zero regardless of the shape. However, I don't think that they have equal/same types due to difference in the shapes and we cannot interchange two empty tensors. Otherwise, we may not be able to perform any other operation later. 

The empty tensor `slice_22: f32[1, 32, 29, 0]` is concatenated with another tensor `add_2: f32[1, 32, 29, 96]` in the above graph. If we interchange one empty tensor with another empty tensor with incompatible shape then concatOp will fail.

**Comment by azecevic000** [comment]
For `concat`, an empty tensor will be neutral to the operation, so I would advise to have a folding pattern that would fold `concat(x1, x2,..., xN, e)` to `concat(x1, x2,..., xN)`, where `e` is neutral. Empty tensor would probably require special handling in both our runtime and TTNN (even if does work for `concat` op, I remember that I saw some cases if TTNN where it would result in an error), so we should try to keep it out of the runtime and opportunistically introduce foldings for ops like the mentioned one when we spot such pattern.

**Comment by mmanzoorTT** [comment]
> For `concat`, an empty tensor will be neutral to the operation, so I would advise to have a folding pattern that would fold `concat(x1, x2,..., xN, e)` to `concat(x1, x2,..., xN)`, where `e` is neutral. Empty tensor would probably require special handling in both our runtime and TTNN (even if does work for `concat` op, I remember that I saw some cases if TTNN where it would result in an error), so we should try to keep it out of the runtime and opportunistically introduce foldings for ops like the mentioned one when we spot such pattern.

ConcatOp is already folded in stablehlo for the above example; so it won't cause any problem. I agree that we should fold such patterns. I will add folding pattern for ConcatOp in another PR.

**Comment by mmanzoorTT** [comment]
@azecevicTT Runtime (#4238) and TTRT (#4234) changes are support empty tensor are checked-in. I also have created an issue #4275 for ConcatOp folding as you suggested.

**Review by azecevicTT — APPROVED** [review]
With fixes in runtime and fold on `ConcatOp` this check makes more sense now.
